structure/redblacktree/rbtree.go

```diff
@@ -0,0 +1,404 @@
+	for _, a := range p {
+		fmt.Printf("%d %v\n", a.Val, a.isRed)
+	}
+}
+
+// Returns inorder representation
+func inorder(currNode *Node) []*Node {
+	var ret = []*Node{}
+	if currNode == nil {
+		return ret
+	}
+
+	ret = append(ret, inorder(currNode.left)...)
+	ret = append(ret, currNode)
+	ret = append(ret, inorder(currNode.right)...)
+	return ret
+}
+
+// Returns preorder representation
+func preorder(currNode *Node) []*Node {
+	var ret = []*Node{}
+	if currNode == nil {
+		return ret
+	}
+	ret = append(ret, currNode)
+	ret = append(ret, preorder(currNode.left)...)
+	ret = append(ret, preorder(currNode.right)...)
+	return ret
+}
+
+// Public search function based on value
+func (t *RBTree) Search(val int) *Node {
+	return search(t.Root, val)
+}
+
+// Internal search function, recursive search
+func search(node *Node, val int) *Node {
+	if node == nil {
+		return nil
+	}
+	if node.Val == val {
+		return node
+	}
+	if val < node.Val {
+		return search(node.left, val)
+	}
+	return search(node.right, val)
+}
+
+// Replaces the node curr with repl
+func (t *RBTree) replace(curr *Node, repl *Node) {
+	if curr == nil {
+		return
+	}
+	if t.isRoot(curr) {
+		t.setRoot(repl)
+		return
+	}
+
+	if curr == curr.parent.left {
+		curr.parent.left = repl
+	} else {
+		curr.parent.right = repl
+	}
+
+	if repl != nil {
+		repl.parent = curr.parent
+	}
+}
+
+// Simple function to test if curr is or should be root
+func (t *RBTree) isRoot(curr *Node) bool {
+	return curr != nil && curr.parent == nil
+}
+
+// Set root (coloring it black is important)
+func (t *RBTree) setRoot(curr *Node) {
+	t.Root = curr
+	if t.Root != nil {
+		t.Root.isRed = false
+	}
+}
+
+// Delete function based on value
+func (t *RBTree) Delete(val int) {
+	var curr = t.Search(val)
+	t.delete(curr)
+}
+
+// Internal delete based on node
+func (t *RBTree) delete(curr *Node) {
+	if curr == nil {
+		// Didn't find the node
+		return
+	}
+
+	var checkIsRed = curr.isRed
+	var next *Node
+
+	// First replace the relevant child, moving it up the list
+	if curr.left == nil {
+		// Only right child/no children
+		next = curr.right
+		t.replace(curr, curr.right)
+	} else if curr.right == nil {
+		// Only left child
+		next = curr.left
+		t.replace(curr, curr.left)
+	} else {
+		// Have to handle both children
+		var minNode = t.min(curr.right)
+		checkIsRed = minNode.isRed
+		next = minNode.right
+		if minNode.parent != curr {
+			t.replace(minNode, minNode.right)
+			minNode.right = curr.right
+			minNode.right.parent = minNode
+		}
+		t.replace(curr, minNode)
+		minNode.left = curr.left
+		minNode.left.parent = minNode
+		minNode.isRed = curr.isRed
+	}
+	if !checkIsRed {
+		// Certain fix up is needed
+		t.fix(next)
+	}
+}
+
+func (t *RBTree) min(curr *Node) *Node {
+	if curr == nil {
+		return nil
+	}
+	for curr.left != nil {
+		curr = curr.left
+	}
+	return curr
+}
+
+func (t *RBTree) fix(curr *Node) {
+	if curr == nil {
+		// Nothing to fix
+		return
+	}
+	for !t.isRoot(curr) && !curr.isRed {
```

Comment:
Try to break this down into different bits because again, because of the sheer number of if statements it has taken me more than 40mins to understand this.

Reply:
Will look at this post-November